lib/nerves_livebook/wifi_monitor.ex

```diff
@@ -0,0 +1,123 @@
+defmodule NervesLivebook.WiFiMonitor do
+  @moduledoc """
+  Monitor WiFi on startup to determine if AP mode needs to be enabled
+
+  This is mostly for devices that must have a WiFi connection in order to interact and
+  make firmware changes. It checks:
+
+  * If any WiFi interface avaialble to configure
+  * If any WiFi interface is configured
+  * If any WiFi interface is connected
+
+  If  WiFi interface is configured, but not connected, then this monitor will watch
+  for a successful connection message for 3 minutes. If no connection is established
+  after that time, then it will set the WiFi interface to AP mode to allow the user
+  to connect directly to the device
+  """
+  use GenServer
+
+  require Logger
+
+  @connected_status [:lan, :internet]
+
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(opts \\ []), do: GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+
+  @impl GenServer
+  def init(_opts) do
+    {:ok, %{wlans: []}, {:continue, :check_for_interfaces}}
+  end
+
+  @impl GenServer
+  def handle_continue(:check_for_interfaces, state) do
+    case all_wifi_interfaces() do
+      [] ->
+        # There are no WiFi interfaces on this device that can be configured, so
+        # skip trying to do anything else.
+        {:noreply, state, :hibernate}
```

Comment:
I think that we need to assume that if the WiFiMonitor is started that there will eventually be a WiFi interface on the device even if there's not one now. The reason is that sometimes USB WiFi adapters take some time to enumerate. It's probably not an issue now since Livebook takes long enough to load that it's probably ok, but on an faster processor, I bet this code could beat enumeration.

Reply:
I changed this to watch for interface presence events and run checks again when a message comes in